# Import Libraries

In [44]:
import pandas as pd
import numpy as np
import os

# Import Data

In [45]:
# Import dataframes
path = r'/Users/hollyringsak/01-2024 Instacart Basket Analysis'

In [46]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [47]:
df_ords_prods_merge.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0_y  order_id  user_id  number_of_orders  \
0             19     5.8          1987   3139998      138                28   
1             19     5.8          1989   1977647      138                30   
2             19     5.8         11433    389851      709                 2   
3             19     5.8         12198    652770      764                 1   
4             19     5.8         12200   1813452      764                 3   

   orders_day_of_week  order_hour_of_day  days_since_last_order  \
0                   6                 11                    3.0   
1                   6                 17                   20.0   
2                   0                 21                    6.0   
3                   3                 13                    NaN   
4                   4                 17                    9.0   

   add_to_cart_order  reordered _merge merge_2  
0                  5          0   both    both  
1                  1          1   both    both  
2                 20          0   both    both  
3                 10          0   both    both  
4                 11          1   both    both

In [48]:
# Create a subset
df = df_ords_prods_merge[:1000000]

In [49]:
# Define a function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [50]:
# Apply the function
df['price_range'] = df.apply(price_label, axis=1)

/tmp/ipykernel_44968/902492192.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [51]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    659373
Low-range product    331283
High range             9344
Name: count, dtype: int64

In [52]:
# Finding max value
df['prices'].max()

24.5

In [53]:
# If-statements with loc() function
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/tmp/ipykernel_44968/639722169.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [54]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [55]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [56]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     659373
Low-range product     331283
High-range product      9344
Name: count, dtype: int64

In [57]:
# if-statements with loc() function on df_ords_prods_merge
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [58]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [59]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [60]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21890146
Low-range product     10126384
High-range product      417682
Name: count, dtype: int64

In [61]:
# If statement with for loops

In [62]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6209632
1    5665830
6    4500246
2    4217766
5    4209449
3    3844096
4    3787193
Name: count, dtype: int64

In [63]:
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [64]:
# Testing results
df_ords_prods_merge['busiest_day']= result

In [65]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22437387
Busiest day        6209632
Least busy         3787193
Name: count, dtype: int64

# Task 4.7

In [66]:
# Finding the 2 busiest days and the 2 slowest days

In [87]:
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result.append("Busiest days")
  elif value == 3 or value ==4:
    result.append("Least busy days")
  else:
    result.append("Regularly busy")


In [88]:
df_ords_prods_merge['busiest_days'] = result

In [89]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy     12927461
Busiest days       11875462
Least busy days     7631289
Name: count, dtype: int64

# The two busiest days combined have a total of 11,875,462 sales, and the two slowest days combined have a total of 7,631,289 sales. 

In [90]:
# Identifying busiest hours of the day

In [91]:
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2764390
11    2738585
14    2691563
15    2664522
13    2663272
12    2620800
16    2537469
9     2456661
17    2089452
8     1719952
18    1637922
19    1259382
20     977017
7      891928
21     796362
22     634737
23     402612
6      290770
0      218942
1      115786
5       88057
2       69431
4       53283
3       51317
Name: count, dtype: int64

In [92]:
result = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value == 10 or value == 11 or value ==14 or value ==15 or value ==13 or value ==12 or value ==16 or value ==9:
    result.append("Most orders")
  elif value ==3 or value ==4 or value==2 or value ==5 or value==1 or value ==0 or value ==6 or value ==23:
    result.append("Fewest orders")
  else:
    result.append("Average orders")

In [93]:
result

['Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most ord

In [95]:
df_ords_prods_merge['busiest_period_of_day'] = result

In [96]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21137262
Average orders    10006752
Fewest orders      1290198
Name: count, dtype: int64

In [97]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge2.pkl'))